static/js/compose_state.js

```diff
@@ -37,13 +38,25 @@ function get_or_set(fieldname, keep_leading_whitespace) {
 }
 
 export function stream_name() {
-    return $("#stream_message_recipient_stream").val().trim();
+    const $dropdown_button = $("#compose_select_stream_name");
+    if (!$dropdown_button) {
+        return "";
+    }
+    const name = $dropdown_button.text().trim();
+    if (name === "Select a stream") {
+        return "";
+    }
+    return name;
 }
 
 export function set_stream_name(newval) {
     if (newval !== undefined) {
-        const $elem = $("#stream_message_recipient_stream");
-        $elem.val(newval);
+        const $dropdown_button = $("#compose_select_stream_name");
+        if (!$dropdown_button) {
+            return;
+        }
+        $dropdown_button.text(newval);
+        compose_ui.on_compose_select_stream_update();
```

Comment:
Importing `compose_ui` here reintroduces an import cycle that I think will prevent us from migrating this module to TypeScript.

Probably the simplest answer is to have the callers of this function be responsible for also calling `compose_ui.on_compose_select_stream_update`, and document that in a comment here.

(This could possibly be done as a preparatory refactor commit, with `compose_ui.on_compose_select_stream_update` being basically a noop function).

The other reasonable option would be to interface with this widget via a new module like we do with `compose_pm_pill`, but I'm not sure if that'll actually avoid an import cycle problem.

Reply:
It feels complicated / confusing to those working in this code in the future, to have callers of this function be responsible for calling the other update functions.

How many import cycles are already present in the code? Could it make more sense to address this issue when thinking about how we want to avoid import cycles more generally?

Reply:
Our current import cycle problem is [gigantic](https://chat.zulip.org/#narrow/stream/6-frontend/topic/cyclical.20dependencies/near/1140366) and not getting better fast enough. On the other hand, import cycles do cause real bugs (examples: [1](https://chat.zulip.org/#narrow/stream/6-frontend/topic/all.20messages/near/1147296), [2](https://chat.zulip.org/#narrow/stream/6-frontend/topic/cyclical.20dependencies/near/1165168), [3](https://chat.zulip.org/#narrow/stream/9-issues/topic/.E2.9C.94.20error.20in.20http.3A.2F.2Flocalhost.3A9991.2Fdevtools.2Fintegrations.2F/near/1285725), [4](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Importing.20user_status.20in.20people.2Ejs.20causes.20invalid.20CSRF.20token/near/1326930)) in addition to the maintainability headaches, and we should try pretty hard to avoid making the problem worse.